library/alloc/src/vec/spec_extend.rs

```diff
@@ -57,31 +100,22 @@ where
     }
 }
 
-impl<T, A: Allocator> SpecExtend<T, IntoIter<T>> for Vec<T, A> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
-        }
-        iterator.ptr = iterator.end;
-    }
-}
-
-impl<'a, T: 'a, I, A: Allocator + 'a> SpecExtend<&'a T, I> for Vec<T, A>
+impl<T, I, A: Allocator> SpecExtendInner<T, I> for Vec<T, A>
 where
-    I: Iterator<Item = &'a T>,
-    T: Clone,
+    I: TrustedLen<Item = T> + TrustedRandomAccess,
 {
-    default fn spec_extend(&mut self, iterator: I) {
-        self.spec_extend(iterator.cloned())
-    }
-}
+    default fn spec_extend(&mut self, mut iterator: I) {
+        let size = iterator.size();
+        self.reserve(size);
 
-impl<'a, T: 'a, A: Allocator + 'a> SpecExtend<&'a T, slice::Iter<'a, T>> for Vec<T, A>
-where
-    T: Copy,
-{
-    fn spec_extend(&mut self, iterator: slice::Iter<'a, T>) {
-        let slice = iterator.as_slice();
-        unsafe { self.append_elements(slice) };
+        // SAFETY: reserve ensured that there is sufficient capacity for the additional items.
+        // The loop upholds the TRA requirements by accessing each element only once.
+        unsafe {
+            let sink = self.as_mut_ptr().add(self.len());
+            for i in 0..size {
+                ptr::write(sink.add(i), iterator.__iterator_get_unchecked(i));
+                self.set_len(self.len() + 1);
+            }
```

Comment:
Is there a reason this loop doesn't use the same pattern as above (i.e., SetLenOnDrop, ptr = ptr.offset(1)...)?

Reply:
`SetLenOnDrop` is a performance optimization to make it easier for the optimizer to realize that the length field doesn't alias during iteration that didn't seem to be necessary here to get results, but we can try adding it and do another perf run.

Reply:
No, this seems okay. Maybe a follow-up could experiment with it but doesn't seem necessary.